x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportInternalInferModelAction.java

```diff
@@ -242,7 +243,13 @@ private void inferAgainstAllocatedModel(
 
         // Get a list of nodes to send the requests to and the number of
         // documents for each node.
-        var nodes = assignment.selectRandomStartedNodesWeighedOnAllocationsForNRequests(request.numberOfDocuments());
+        var nodes = assignment.selectRandomStartedNodesWeighedOnAllocationsForNRequests(request.numberOfDocuments(), RoutingState.STARTED);
+
+        // We couldn't find any nodes in the started state so let's look for ones that are stopping in case we're shutting down some nodes
```

Comment:
I think what's done in this PR is an improvement on what we have today in terms of not failing requests during rolling upgrades.

For a followup that might help even more I'm thinking we could route requests to nodes where the routing state is `STARTING`. We could build up a queue of items to send to the C++ process before it's ready, then feed them through once it's running. We'd have to decide whether to favour `STOPPING` or `STARTING` after failing to find a `STARTED`, and there would probably be work elsewhere to accept requests into the queue before the C++ process is ready. But it could be another layer of improvement even if we still haven't reached perfection.

Reply:
Sounds good!